keps/sig-cli/2551-return-code-normalization/README.md

```diff
@@ -0,0 +1,637 @@
+for the developers to warn when a new deployment fails because of the lack of some permission, so those permissions can be 
+updated for the pipeline to work correctly.
+
+The developers are making a lot of changes, and they keeps asking for Bruce to look for every pipeline execution, even those that 
+fails because of wrong manifests and not because of authorization issues. So Bruce needs a new mechanism: That the pipeline knows
+when it fails because of the lack of the authorization on Kubernetes API Server, and then the warning is sent to the security team
+only when the pipeline breaks because of a specific error code that represents this authorization failure.
+
+#### Story 3
+Roberta works as the product manager of a big CI/CD SaaS provider. They want to have in their marketplace the
+execution of kubernetes commands targeting a cluster, and fastly providing a feedback to the user if the error
+was due to something on the client side (like a missing flag, an invalid yaml file because of...tabs...) or if this
+is due to some invalid operation on the server side.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+* Users already relying on specific error codes might face migration problems / false positives or negatives 
+because of some specific error code change. The proposal here is to decide if the old or new return code behavior
+will be used based on a well defined environment variable. The old behavior will be kept and followed by the 
+deprecation path will be defaulted after a number of releases.
+
+#### Feature Gating
+In an effort to roll out these changes slowly and prevent flag bloat, this feature will be enabled using the environment variable `KUBECTL_ERROR_CODES`, which will default to an empty value.
+To enable the feature, the value must be non-empty, and not equal `0`, or `false`.
+
+## Design Details
+
+The majority of commands already are organized as the following:
+* Validation: check `Options` and set defaults. This runs on the client side.
+* Execute based on `Options`. This might run on the client side, be dry-run, run an external command or call the APIServer.
+
+One thing that should be done is map all the error codes/ints as constants in some file, so they can be automatically
+documented.
+
+### Error Codes
+The following table represents the proposed error codes and the condition describing its usage.
+
+| Code  | Description                                                                                           |
+| ----- | ----------------------------------------------------------------------------------------------------- |
+| 1-200 | Reserved for exit codes from exec and external commands                                               |
```

Comment:
I was wondering if we need to reserve all 200, if just the initial 100 isn't sufficient? But that's something we'll need to empirically figure out as we go.

Reply:
agreed. This was a discussion I was on with Ross, that maybe the code reservation can be something more sparse or in a "range" other than reserving all the 200

Reply:
Let's "close" this list before going to beta, for alpha let's leave it open.